### PR TITLE
[Chat] Add KTO

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by TongLi3701** [comment]
@YeAnbang Seem that you need to rename testing data.

**Review by TongLi3701 — COMMENTED** [review]
Thanks Anbang.

Some comments.

File: applications/ColossalChat/coati/models/lora.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def __init__(
         self.fan_in_fan_out = fan_in_fan_out
         # Actual trainable parameters
         if r > 0:
-            self.lora_A = nn.Parameter(self.weight.new_zeros((r, in_features)))
+            self.lora_A = nn.Parameter(self.weight.new_zeros((r, in_features)), requires_grad=False)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/lora.py:53]: I was confused about this originally. you are right, should be true

File: applications/ColossalChat/coati/dataset/loader.py [hunk 2]
```diff
@@ -235,6 +235,91 @@ def __call__(self, instances: Sequence[Dict[str, List[int]]]) -> Dict[str, torch
...
+        loss_mask = torch.nn.utils.rnn.pad_sequence(
+            sequences=loss_mask, batch_first=True, padding_value=0
+        )  # (bsz, max_len)
+        to_pad = self.max_length - input_ids.size(1)
+        input_ids = F.pad(input_ids, (0, to_pad), value=self.tokenizer.pad_token_id)
+        loss_mask = F.pad(loss_mask, (0, to_pad), value=0)
+        attention_mask = input_ids.ne(self.tokenizer.pad_token_id)  # `torch.BoolTensor`, (bsz, max_len)
+
+        # prepare kt data
+        kl_completion = completion[::-1]  # y'
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/loader.py:294]: no, this is to create irrelevant data pairs for calculating the KL part of the KTO loss, this particular line reverse the order of completion to create irrelevant data pairs.

File: applications/ColossalChat/coati/trainer/kto.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+                    input_ids=kl_input_ids,
+                    attention_mask=kl_attention_mask,
+                )["logits"]
+
+            logits = self.model(
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+            )["logits"]
+
+            logprob = calc_masked_log_probs(logits, input_ids, loss_mask[:, 1:]).sum(-1)
```
Comment (TongLi3701) [inline applications/ColossalChat/coati/trainer/kto.py:152]: Seem that this part is wrong?
https://github.com/hpcaitech/ColossalAI/blob/9688e19b322510bd64956a75bd6227850817cc69/applications/ColossalChat/coati/models/utils.py#L107

Comment (TongLi3701) [reply applications/ColossalChat/coati/trainer/kto.py:152]: -1 does not shift left by one ?

File: applications/ColossalChat/coati/dataset/utils.py [hunk 4]
```diff
@@ -106,10 +119,17 @@ def tokenize_and_concatenate(tokenizer: PreTrainedTokenizer, text: List[str], re
...
-            loss_starts.append(len(input_ids))
-            loss_ends.append(len(input_ids) + len(tokenized))
-        input_ids.extend(tokenized)
+        if len(input_ids) + len(tokenized) <= max_length or len(loss_ends) == 0:
+            if r:
+                loss_starts.append(len(input_ids))
+                loss_ends.append(len(input_ids) + len(tokenized))
+            input_ids.extend(tokenized)
+    if loss_starts[0] >= max_length:
+        return None, None, None
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/utils.py:128]: Raising an error breaks the pipeline and in a dataset of a certain scale, there must be some data samples whose first user query is longer than the max length. Currently, we just ignore those.

File: applications/ColossalChat/examples/training_scripts/train_kto.py [hunk 5]
```diff
@@ -117,18 +116,15 @@ def train(args):
         else:
             model = AutoModelForCausalLM.from_pretrained(args.pretrain)
         disable_dropout(model)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/training_scripts/train_kto.py:118]: no, for rl methods, dropout is always disabled to ensure stable training

File: applications/ColossalChat/coati/trainer/kto.py (hunk 3 above)
Comment (YeAnbang) [reply applications/ColossalChat/coati/trainer/kto.py:152]: this part calculates the log probability of each generated tokens. As the logits at position i correspond to prob of label[i+1] being generated, the code is correct. Here is the corresponding implementation from TRL.
image

File: applications/ColossalChat/examples/data_preparation_scripts/prepare_dataset.py [hunk 6]
```diff
@@ -228,10 +236,13 @@ def main():
...
-        dataset = dataset.filter(
-            lambda data: data["chosen_input_ids" if args.type == "preference" else "input_ids"] is not None
-        )
+        if args.type == "kto":
+            filter_by = "completion"
+        elif args.type == "preference":
+            filter_by = "chosen_input_ids"
+        else:
+            filter_by = "input_ids"
+        dataset = dataset.filter(lambda data: data[filter_by] is not None)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/data_preparation_scripts/prepare_dataset.py:245]: remove overlength data samples, the tokenization function will return None

**Review by TongLi3701 — APPROVED** [review]
Thanks Anbang.

Overall looks good. Just some minor questions.

File: applications/ColossalChat/benchmarks/prepare_dummy_test_dataset.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            args.dataset_size,
+        )
+    elif args.data_type == "kto":
+        dataset = DummyLLMDataset(
+            ["prompt", "completion", "label"],
+            args.max_length - 512,
+            args.dataset_size,
+            gen_fn={
+                "completion": lambda x: [1] * 512,
+                "label": lambda x: x % 2,
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/benchmarks/prepare_dummy_test_dataset.py:64]: in the benchmark, context is set to 2048-512, completion is set to 512 tokens

File: applications/ColossalChat/coati/dataset/tokenization_utils.py [hunk 8]
```diff
@@ -171,60 +126,50 @@ def supervised_tokenize_sft(
...
-        target_turn = target_turn - 1
+        template.messages = template.messages[:-1]
 
     # Prepare data
-    prompt = template.get_prompt(target_turn, add_generation_prompt=True)
-    chunks, require_loss = split_templated_prompt_into_chunks(
-        template.messages[:target_turn], prompt, conversation_template.end_of_assistant
-    )
-    tokenized, starts, ends = tokenize_and_concatenate(tokenizer, chunks, require_loss)
+    prompt = template.get_prompt(length=len(template.messages) - 1, add_generation_prompt=True)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/tokenization_utils.py:165]: this is the context hence it doesn't include the last line from assistance.

File: applications/ColossalChat/coati/dataset/tokenization_utils.py [hunk 9]
```diff
@@ -405,3 +319,62 @@ def tokenize_rlhf(
...
+        if mess.get("from", None) == "user":
+            template.append_message("user", mess["content"])
+        elif mess.get("from", None) == "assistant":
+            template.append_message("assistant", mess["content"])
+        else:
+            raise ValueError(f"Unsupported role {mess.get('from', None)}")
+    generation_prompt = template.get_prompt(len(prompt), add_generation_prompt=True)
+    template.append_message("assistant", completion["content"])
+    full_prompt = template.get_prompt(len(prompt) + 1, add_generation_prompt=False)
+    tokenized_full_prompt = tokenizer(full_prompt, add_special_tokens=False)["input_ids"]
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/tokenization_utils.py:363]: yes, we can. The add_generation_prompt=True flag are for inference and should be identical to the context part of the full prompt
